### PR TITLE
Rewrote the Parser

### PR DESCRIPTION
**Pull request by azreika**
Completely rewrote the parser.
* Refactored the rules globally so that structure + information passed make more sense
* Several hidden bugs fixed 🐛🔨, particularly rules mixing up where Souffle multi-name identifiers should be used (e.g. `comp1.comp2.comp3.nestedthingiwant`) vs. just single-name identifiers (e.g. `myrel`)
* Changed from number use to name use in rule semantics. E.g. for the syntactic rule, `arg : COUNT COLON atom`, the element `atom` is now referred to using `$atom` instead of `$3`. This is done to increase readability and make parser prototyping a lot less cumbersome when altering rules.
* Typecasts now work correctly (using changes made by @olligobber)

A big reason for this refactoring was to clean up the parser so that future feature additions are possible without causing issues, namely the changes required for the new typesystem (e.g. adding record identifiers to record constructors). The changes can now be made without introducing several reduce/reduce & shift/reduce conflicts 🎉. The final changes should be added in the next major version of Souffle as they may break current programs: changing `.` to `::` for component scoping, and prepending record type identifiers to record constructors. These changes are marked with TODOs in the parser.

### CODE REVIEW

File: src/parser.yy [hunk 1]
```diff
@@ -162,411 +153,542 @@
...
-%destructor { for (auto* cur : $$) { delete cur; } } head rule rule_def iodirective_list load_head store_head relation_list relation_decl
-%destructor { delete $$; } body conjunction disjunction literal term
-%destructor { delete $$; } rel_id
-%destructor { delete $$; } attributes non_empty_attributes
-
-%printer { yyoutput << $$; } <*>;
-
-%precedence AS
+/* -- Non-Terminal Types -- */
+%type <std::vector<std::string>>        identifier
```
Comment (mmcgr) [inline src/parser.yy]: Is there an ordering to this list? I don't immediately see it. If there is not, impose one to help with readability. (lexicographical or semantic)

File: src/parser.yy [hunk 2]
```diff
@@ -162,411 +153,542 @@
...
 /* Program */
 program
   : unit
+  ;
 
 /* Top-level statement */
 unit
   : unit type {
-        driver.addType(std::unique_ptr<AstType>($2));
+        driver.addType(std::unique_ptr<AstType>($type));
```
Comment (mmcgr) [inline src/parser.yy]: I'm really like the change to names instead of index.

**Comment by mmcgr** [comment]
This seems to have introduced some memory leaks in standard running of souffle. e.g., the 'cat' test finds leaks in objects made in parser.yy on lines 717, 603, 620, 622. src/test/ast_program_test and src/test/ast_utils_test also fail due to parser leaks of objects created on those lines.

Note that this is in standard running, not after parsing errors.

**Comment by azreika** [comment]
These memory leaks have been fixed by PR #1010.

**Pull request merged**